packages/tpetra/core/src/Tpetra_MultiVector_def.hpp

```diff
@@ -3395,15 +3395,23 @@ namespace Tpetra {
     // be exactly what we need, so we won't have to resize them later.
     {
       const size_t newSize = X.imports_.extent (0) / numCols;
+      const size_t offset = j*newSize;
```

Comment:
`j` isn't necessarily the correct offset.  We should actually use `jj` (see above) here, in case `*this*` views a noncontiguous set of columns of another MultiVector.

Actually, if Y views a contiguous proper subset of columns of X, and Z views a noncontiguous proper subset of columns of Y, then `jj` is not even the right entry of `imports_`.  We can fix that in the "view contiguous subset of columns" constructors and methods by adjusting the offset of `imports_` and `exports_` accordingly.

Reply:
So I should switch j to jj?  Note that previously, the subviews took the first newSize entries of the buffer always.  If the buffers are strictly used for communication, that might be fine, but I thought indexing into them would be better.

Reply:
@kddevin wrote:
>  If the buffers are strictly used for communication, ....

That's correct.  They are used for nothing other than communication of packed data.

> So I should switch `j` to `jj`?

Yes.  I actually studied the example in my above comments, and I think this is a complete fix, given how `MultiVector::subView*` work. 

If we switch `j` to `jj`, that will fix the case where different single Vectors view distinct columns of one MultiVector.  In this case, your change will mean that different Vectors will view different parts of the buffer.  Thus, you'll be able to do concurrent asynchronous doExport or doImport correctly with the Vectors.

Consider the following, more complicated example:
```cpp
using MV = Tpetra::MultiVector<>;
MV X (map, 10);
Teuchos::Range1D Y_colRange (1, 5); // inclusive: 1, 2, 3, 4, 5
RCP<MV> Y = X.subViewNonConst (Y_colRange);

Teuchos::Array<size_t> whichVecsZ {{0, 2, 4}};
Teuchos::Array<size_t> whichVecsW {{0, 2, 4}};
RCP<MV> Z = X.subViewNonConst (whichVecsZ ());
RCP<MV> W = Y.subViewNonConst (whichVecsW ());
```
Y views columns 1, 2, 3, 4, and 5 of X.  Z views columns 0, 2, and 4 of X.  W views columns 0, 2, and 4 of Y -- thus, columns 1, 3, and 5 of X.  Users would expect it to be safe to do concurrent communication with Z and W.

In this case, they are right, because neither overload of `subViewNonConst` reuses their parent MultiVector's buffers.  Thus, Z and W each would start with zero-length `imports_` and `exports_` buffers, and would each reallocate them on demand.

Thus, this case is already correct, and your change doesn't break it.  If we decide to make `subView*` reuse communication buffers in the future, we only need to use the correct offset.  That will make the above example continue to be correct.

Reply:
The only way we could write a reliable test for this use case, though, would be if we had asynchronous doExport / doImport.  We don't have that right now.